<a href="https://colab.research.google.com/github/glaiml/prj12-nlp-sarcasm-detection/blob/master/VGK_NLP_Project_Sarcasm_Detection_Questions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sarcasm Detection
 **Acknowledgement**

Misra, Rishabh, and Prahal Arora. "Sarcasm Detection using Hybrid Neural Network." arXiv preprint arXiv:1908.07414 (2019).

**Required Files given in below link.**

https://drive.google.com/drive/folders/1xUnF35naPGU63xwRDVGc-DkZ3M8V5mMk

## Install `Tensorflow2.0` 

In [1]:
!!pip uninstall tensorflow
!pip install tensorflow==2.0.0

     |████████████████████████████████| 86.3MB 49kB/s 
     |████████████████████████████████| 450kB 59.9MB/s 
     |████████████████████████████████| 3.8MB 63.3MB/s 
  Created wheel for gast: filename=gast-0.2.2-cp36-none-any.whl size=7540 sha256=2eb1eab696273c82208927bfcac108543c417b3addf33a4f2eb527b90b19ac29
  Stored in directory: /root/.cache/pip/wheels/5c/2e/7e/a1d4d4fcebe6c381f378ce7743a3ced3699feb89bcfbdadadd
Successfully built gast
  Found existing installation: tensorflow-estimator 2.2.0rc0
    Uninstalling tensorflow-estimator-2.2.0rc0:
      Successfully uninstalled tensorflow-estimator-2.2.0rc0
  Found existing installation: tensorboard 2.1.1
    Uninstalling tensorboard-2.1.1:
      Successfully uninstalled tensorboard-2.1.1
  Found existing installation: gast 0.3.3
    Uninstalling gast-0.3.3:
      Successfully uninstalled gast-0.3.3


## Get Required Files from Drive

In [2]:
from google.colab import drive
drive.mount('/content/drive/')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive/


In [0]:
#Set your project path 
import os
os.chdir('/content/drive/My Drive')
project_path =  "AIML-NLP-Data/"

#**## Reading and Exploring Data**

## Read Data "Sarcasm_Headlines_Dataset.json". Explore the data and get  some insights about the data. ( 4 marks)
Hint - As its in json format you need to use pandas.read_json function. Give paraemeter lines = True.

In [0]:
headlines_file = project_path + "Sarcasm_Headlines_Dataset.json"
import pandas as pd
headlines = pd.read_json(headlines_file,lines=True)

In [11]:
headlines.head(5)

article_link  ... is_sarcastic
0  https://www.huffingtonpost.com/entry/versace-b...  ...            0
1  https://www.huffingtonpost.com/entry/roseanne-...  ...            0
2  https://local.theonion.com/mom-starting-to-fea...  ...            1
3  https://politics.theonion.com/boehner-just-wan...  ...            1
4  https://www.huffingtonpost.com/entry/jk-rowlin...  ...            0

[5 rows x 3 columns]

Loading is ok

In [12]:
headlines.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26709 entries, 0 to 26708
Data columns (total 3 columns):
article_link    26709 non-null object
headline        26709 non-null object
is_sarcastic    26709 non-null int64
dtypes: int64(1), object(2)
memory usage: 626.1+ KB


No nulls, 26709 entries

In [13]:
headlines['is_sarcastic'].value_counts()

0    14985
1    11724
Name: is_sarcastic, dtype: int64

There is an imbalance in data, there are more non sarcastic than sarcastic. MAybe that can skew the model? We may need to make it equal

In [0]:
sarcastic_headlines = headlines.loc[headlines['is_sarcastic']==1]
#undersample non sarcastic headlines
non_sarcastic_headlines = headlines.loc[headlines['is_sarcastic']==0].sample(n=11724,random_state=10)

In [27]:
sarcastic_headlines.head()

article_link  ... is_sarcastic
2   https://local.theonion.com/mom-starting-to-fea...  ...            1
3   https://politics.theonion.com/boehner-just-wan...  ...            1
8   https://politics.theonion.com/top-snake-handle...  ...            1
15  https://entertainment.theonion.com/nuclear-bom...  ...            1
16  https://www.theonion.com/cosby-lawyer-asks-why...  ...            1

[5 rows x 3 columns]

In [28]:
sarcastic_headlines.count()

article_link    11724
headline        11724
is_sarcastic    11724
dtype: int64

In [29]:
non_sarcastic_headlines.head()

article_link  ... is_sarcastic
22600  https://www.huffingtonpost.com/entry/5-serious...  ...            0
25792  https://www.huffingtonpost.com/entry/obama-edu...  ...            0
20038  https://www.huffingtonpost.com/entry/majority-...  ...            0
6564   https://www.huffingtonpost.com/entry/neocamp-o...  ...            0
20070  https://www.huffingtonpost.com/entry/donald-tu...  ...            0

[5 rows x 3 columns]

In [30]:
non_sarcastic_headlines.count()

article_link    11724
headline        11724
is_sarcastic    11724
dtype: int64

In [0]:
balanced_headlines=pd.concat([sarcastic_headlines, non_sarcastic_headlines])

In [34]:
balanced_headlines['is_sarcastic'].value_counts()

1    11724
0    11724
Name: is_sarcastic, dtype: int64

## Drop `article_link` from dataset. ( 2 marks)
As we only need headline text data and is_sarcastic column for this project. We can drop artical link column here.

In [0]:
headlines.drop('article_link', axis=1, inplace=True)
balanced_headlines.drop('article_link', axis=1, inplace=True)

In [33]:
balanced_headlines.head()

headline  is_sarcastic
2   mom starting to fear son's web series closest ...             1
3   boehner just wants wife to listen, not come up...             1
8   top snake handler leaves sinking huckabee camp...             1
15  nuclear bomb detonates during rehearsal for 's...             1
16  cosby lawyer asks why accusers didn't come for...             1

## Clean Text?

## Get the Length of each line and find the maximum length. ( 4 marks)
As different lines are of different length. We need to pad the our sequences using the max length.

In [0]:
contractions = { 
"ain't": "am not / are not / is not / has not / have not",
"aren't": "are not / am not",
"can't": "cannot",
"can't've": "cannot have",
"'cause": "because",
"could've": "could have",
"couldn't": "could not",
"couldn't've": "could not have",
"didn't": "did not",
"doesn't": "does not",
"don't": "do not",
"hadn't": "had not",
"hadn't've": "had not have",
"hasn't": "has not",
"haven't": "have not",
"he'd": "he had / he would",
"he'd've": "he would have",
"he'll": "he shall / he will",
"he'll've": "he shall have / he will have",
"he's": "he has / he is",
"how'd": "how did",
"how'd'y": "how do you",
"how'll": "how will",
"how's": "how has / how is / how does",
"I'd": "I had / I would",
"I'd've": "I would have",
"I'll": "I shall / I will",
"I'll've": "I shall have / I will have",
"I'm": "I am",
"I've": "I have",
"isn't": "is not",
"it'd": "it had / it would",
"it'd've": "it would have",
"it'll": "it shall / it will",
"it'll've": "it shall have / it will have",
"it's": "it has / it is",
"let's": "let us",
"ma'am": "madam",
"mayn't": "may not",
"might've": "might have",
"mightn't": "might not",
"mightn't've": "might not have",
"must've": "must have",
"mustn't": "must not",
"mustn't've": "must not have",
"needn't": "need not",
"needn't've": "need not have",
"o'clock": "of the clock",
"oughtn't": "ought not",
"oughtn't've": "ought not have",
"shan't": "shall not",
"sha'n't": "shall not",
"shan't've": "shall not have",
"she'd": "she had / she would",
"she'd've": "she would have",
"she'll": "she shall / she will",
"she'll've": "she shall have / she will have",
"she's": "she has / she is",
"should've": "should have",
"shouldn't": "should not",
"shouldn't've": "should not have",
"so've": "so have",
"so's": "so as / so is",
"that'd": "that would / that had",
"that'd've": "that would have",
"that's": "that has / that is",
"there'd": "there had / there would",
"there'd've": "there would have",
"there's": "there has / there is",
"they'd": "they had / they would",
"they'd've": "they would have",
"they'll": "they shall / they will",
"they'll've": "they shall have / they will have",
"they're": "they are",
"they've": "they have",
"to've": "to have",
"wasn't": "was not",
"we'd": "we had / we would",
"we'd've": "we would have",
"we'll": "we will",
"we'll've": "we will have",
"we're": "we are",
"we've": "we have",
"weren't": "were not",
"what'll": "what shall / what will",
"what'll've": "what shall have / what will have",
"what're": "what are",
"what's": "what has / what is",
"what've": "what have",
"when's": "when has / when is",
"when've": "when have",
"where'd": "where did",
"where's": "where has / where is",
"where've": "where have",
"who'll": "who shall / who will",
"who'll've": "who shall have / who will have",
"who's": "who has / who is",
"who've": "who have",
"why's": "why has / why is",
"why've": "why have",
"will've": "will have",
"won't": "will not",
"won't've": "will not have",
"would've": "would have",
"wouldn't": "would not",
"wouldn't've": "would not have",
"y'all": "you all",
"y'all'd": "you all would",
"y'all'd've": "you all would have",
"y'all're": "you all are",
"y'all've": "you all have",
"you'd": "you had / you would",
"you'd've": "you would have",
"you'll": "you shall / you will",
"you'll've": "you shall have / you will have",
"you're": "you are",
"you've": "you have"
}

In [0]:
def clean_text(text, remove_stopwords=True):
  # Convert words to lower case
  text = text.lower()
  if True:
    text = text.split()
    new_text = []
    for word in text:
      if word in contractions:
        new_text.append(contractions[word])
      else:
        new_text.append(word)
      text = " ".join(new_text)
      text = re.sub(r'https?:\/\/.*[\r\n]*', '', text, flags=re.MULTILINE)
      text = re.sub(r'\<a href', ' ', text)
      text = re.sub(r'&amp;', '', text)
      text = re.sub(r'[_"\-;%()|+&=*%.,!?:#$@\[\]/]', ' ', text)
      text = re.sub(r'<br />', ' ', text)
      text = re.sub(r'\'', ' ', text)
      if remove_stopwords:
        text = text.split()
        stops = set(stopwords.words("english"))
        text = [w for w in text if not w in stops]
        text = " ".join(text)
  return text

In [36]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [56]:
# Clean the headlines
import re
from nltk.corpus import stopwords 
from pickle import dump, load

clean_balanced_headlines = []
for headline in balanced_headlines['headline']:
  clean_balanced_headlines.append(clean_text(headline, remove_stopwords=False))
print("balanced headlines are cleaned.")

balanced headlines are cleaned.


In [0]:
balanced_headlines['cleaned_headline']=clean_balanced_headlines

In [58]:
balanced_headlines.head()

headline  ...                                   cleaned_headline
2   mom starting to fear son's web series closest ...  ...  mom starting to fear son s web series closest ...
3   boehner just wants wife to listen, not come up...  ...  boehner just wants wife to listen  not come up...
8   top snake handler leaves sinking huckabee camp...  ...  top snake handler leaves sinking huckabee camp...
15  nuclear bomb detonates during rehearsal for 's...  ...  nuclear bomb detonates during rehearsal for  s...
16  cosby lawyer asks why accusers didn't come for...  ...  cosby lawyer asks why accusers did not come fo...

[5 rows x 3 columns]

In [0]:
balanced_headlines.drop('headline', axis=1, inplace=True)

In [60]:
balanced_headlines.head()

is_sarcastic                                   cleaned_headline
2              1  mom starting to fear son s web series closest ...
3              1  boehner just wants wife to listen  not come up...
8              1  top snake handler leaves sinking huckabee camp...
15             1  nuclear bomb detonates during rehearsal for  s...
16             1  cosby lawyer asks why accusers did not come fo...

Doing MAX length

In [66]:
print (balanced_headlines['cleaned_headline'][2])

mom starting to fear son s web series closest thing she will have to grandchild


In [70]:
len(balanced_headlines['cleaned_headline'][2].split())

15

In [72]:
max_length = max([len(headline.split()) for headline in balanced_headlines['cleaned_headline']])
print (max_length)

39


#**## Modelling**

## Import required modules required for modelling.

In [0]:
import numpy as np
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.layers import Dense, Input, LSTM, Embedding, Dropout, Activation, Flatten, Bidirectional, GlobalMaxPool1D
from tensorflow.keras.models import Model, Sequential

# Set Different Parameters for the model. ( 2 marks)

In [0]:
max_features = 10000
maxlen = max_length
embedding_size = 200

## Apply Keras Tokenizer of headline column of your data.  ( 4 marks)
Hint - First create a tokenizer instance using Tokenizer(num_words=max_features) 
And then fit this tokenizer instance on your data column df['headline'] using .fit_on_texts()

In [0]:
num_words=max_features
tokenizer = Tokenizer(num_words=max_features)
tokenizer.fit_on_texts(balanced_headlines['cleaned_headline'])

# Define X and y for your model.

In [88]:
X = tokenizer.texts_to_sequences(balanced_headlines['cleaned_headline'])
X = pad_sequences(X, maxlen = maxlen)
y = np.asarray(balanced_headlines['is_sarcastic'])

print("Number of Samples:", len(X))
print(X[0])
print("Number of Labels: ", len(y))
print(y[0])

Number of Samples: 23448
[   0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0  120  766    1  823
  231    4 2203  657 4289  212   72   33   40    1 9650]
Number of Labels:  23448
1


## Get the Vocabulary size ( 2 marks)
Hint : You can use tokenizer.word_index.

In [89]:
print(tokenizer.word_counts)
print(tokenizer.document_count)
print(tokenizer.word_index)
print(tokenizer.word_docs)

OrderedDict([('mom', 200), ('starting', 43), ('to', 7351), ('fear', 41), ('son', 119), ('s', 3962), ('web', 16), ('series', 49), ('closest', 7), ('thing', 129), ('she', 312), ('will', 662), ('have', 504), ('grandchild', 2), ('boehner', 24), ('just', 528), ('wants', 113), ('wife', 86), ('listen', 19), ('not', 1389), ('come', 93), ('up', 758), ('with', 1601), ('alternative', 12), ('debt', 25), ('reduction', 3), ('ideas', 34), ('top', 106), ('snake', 8), ('handler', 4), ('leaves', 59), ('sinking', 6), ('huckabee', 18), ('campaign', 172), ('nuclear', 58), ('bomb', 18), ('detonates', 2), ('during', 195), ('rehearsal', 2), ('for', 2888), ('spider', 12), ('man', 1343), ('musical', 20), ('cosby', 17), ('lawyer', 27), ('asks', 84), ('why', 413), ('accusers', 8), ('did', 141), ('forward', 34), ('be', 731), ('smeared', 1), ('by', 829), ('legal', 29), ('team', 80), ('years', 211), ('ago', 26), ('stock', 25), ('analysts', 3), ('confused', 11), ('frightened', 3), ('boar', 4), ('market', 30), ('court

In [90]:
#ref -- https://stackoverflow.com/questions/53525994/how-to-find-num-words-or-vocabulary-size-of-keras-tokenizer-when-one-is-not-as
vocab_size = len(tokenizer.word_index) + 1
print(vocab_size)

24121


#**## Word Embedding**

## Get Glove Word Embeddings

In [0]:
glove_file = project_path + "glove.6B.zip"

In [0]:
#Extract Glove embedding zip file
from zipfile import ZipFile
with ZipFile(glove_file, 'r') as z:
  z.extractall()

# Get the Word Embeddings using Embedding file as given below.

In [0]:
EMBEDDING_FILE = './glove.6B.200d.txt'

embeddings = {}
for o in open(EMBEDDING_FILE):
    word = o.split(" ")[0]
    # print(word)
    embd = o.split(" ")[1:]
    embd = np.asarray(embd, dtype='float32')
    # print(embd)
    embeddings[word] = embd



# Create a weight matrix for words in training docs

In [98]:
embedding_matrix = np.zeros((vocab_size, 200))

for word, i in tokenizer.word_index.items():
    embedding_vector = embeddings.get(word)
    if embedding_vector is not None:
        embedding_matrix[i] = embedding_vector

len(embeddings.values())

400000

## Create and Compile your Model  ( 7 marks)
Hint - Use Sequential model instance and then add Embedding layer, Bidirectional(LSTM) layer, then dense and dropout layers as required. 
In the end add a final dense layer with sigmoid activation for binary classification.


In [103]:
### create the model
model = Sequential()

### Embedding layer for hint 
model.add(Embedding(vocab_size, embedding_size, weights = [embedding_matrix]))
### Bidirectional LSTM layer for hint 
model.add(Bidirectional(LSTM(512, return_sequences = True)))
### dense
model.add(Dense(256, activation='relu'))
##dropout
model.add(Dropout(0.2, input_shape=(256,)))
##sigmoid
model.add(Dense(1, activation='sigmoid'))
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_2 (Embedding)      (None, None, 200)         4824200   
_________________________________________________________________
bidirectional_2 (Bidirection (None, None, 1024)        2920448   
_________________________________________________________________
dense_4 (Dense)              (None, None, 256)         262400    
_________________________________________________________________
dropout_2 (Dropout)          (None, None, 256)         0         
_________________________________________________________________
dense_5 (Dense)              (None, None, 1)           257       
Total params: 8,007,305
Trainable params: 8,007,305
Non-trainable params: 0
_________________________________________________________________


# Fit your model with a batch size of 100 and validation_split = 0.2. and state the validation accuracy ( 5 marks)


In [104]:
batch_size = 100
epochs = 5

sarcasm_detector = model.fit(X, y, batch_size=batch_size, epochs=epochs, verbose=1,validation_split=0.2)

Train on 18758 samples, validate on 4690 samples
Epoch 1/5
18758/18758 [==============================] - 374s 20ms/sample - loss: 0.4525 - accuracy: 0.7897 - val_loss: 0.5182 - val_accuracy: 0.7794
Epoch 2/5
18758/18758 [==============================] - 367s 20ms/sample - loss: 0.2773 - accuracy: 0.8887 - val_loss: 0.3316 - val_accuracy: 0.8639
Epoch 3/5
18758/18758 [==============================] - 384s 20ms/sample - loss: 0.1898 - accuracy: 0.9301 - val_loss: 0.5505 - val_accuracy: 0.7560
Epoch 4/5
18758/18758 [==============================] - 379s 20ms/sample - loss: 0.1277 - accuracy: 0.9549 - val_loss: 0.7995 - val_accuracy: 0.7611
Epoch 5/5
18758/18758 [==============================] - 377s 20ms/sample - loss: 0.0830 - accuracy: 0.9725 - val_loss: 0.7148 - val_accuracy: 0.8156


## Summary

1. We have read article headlines from huffpost and onion
2. Balanced the data (Since there were lesser sarcastic news)
3. Cleaned the data
4. Generated word embeddings using glove 200d
5. Built a simple bidirectional LSTM Model to detect and predict sarcastic news

## Overall Validation accuracy of the model is at 81.56 %